core/multivm_processor.go

```diff
@@ -0,0 +1,163 @@
+			} else {
+				vm.CommitNonexist(address)
+			}
+		case sputnikvm.RequireAccountCode:
+			address := ret.Address()
+			if statedb.Exist(address) {
+				vm.CommitAccountCode(address, statedb.GetCode(address))
+			} else {
+				vm.CommitNonexist(address)
+			}
+		case sputnikvm.RequireAccountStorage:
+			address := ret.Address()
+			key := common.BigToHash(ret.StorageKey())
+			if statedb.Exist(address) {
+				value := statedb.GetState(address, key).Big()
+				key := ret.StorageKey()
+				vm.CommitAccountStorage(address, key, value)
+			} else {
+				vm.CommitNonexist(address)
+			}
+		case sputnikvm.RequireBlockhash:
+			number := ret.BlockNumber()
+			hash := bc.GetBlockByNumber(number.Uint64()).Hash()
+			vm.CommitBlockhash(number, hash)
+		}
+	}
+
+	// VM execution is finished at this point. We apply changes to the statedb.
+
+	for _, account := range vm.AccountChanges() {
+		switch account.Typ() {
+		case sputnikvm.AccountChangeIncreaseBalance:
+			address := account.Address()
+			amount := account.ChangedAmount()
+			statedb.AddBalance(address, amount)
+		case sputnikvm.AccountChangeDecreaseBalance:
+			address := account.Address()
+			amount := account.ChangedAmount()
+			balance := new(big.Int).Sub(statedb.GetBalance(address), amount)
+			statedb.SetBalance(address, balance)
+		case sputnikvm.AccountChangeRemoved:
+			address := account.Address()
+			statedb.Suicide(address)
+		case sputnikvm.AccountChangeFull:
+			address := account.Address()
+			code := account.Code()
+			nonce := account.Nonce()
+			balance := account.Balance()
+			statedb.SetBalance(address, balance)
+			statedb.SetNonce(address, nonce.Uint64())
+			statedb.SetCode(address, code)
+			for _, item := range account.ChangedStorage() {
+				statedb.SetState(address, common.BigToHash(item.Key), common.BigToHash(item.Value))
+			}
+		case sputnikvm.AccountChangeCreate:
+			address := account.Address()
+			code := account.Code()
+			nonce := account.Nonce()
+			balance := account.Balance()
+			statedb.SetBalance(address, balance)
+			statedb.SetNonce(address, nonce.Uint64())
+			statedb.SetCode(address, code)
+			for _, item := range account.Storage() {
+				statedb.SetState(address, common.BigToHash(item.Key), common.BigToHash(item.Value))
+			}
+		default:
+			panic("unreachable")
+		}
+	}
+	for _, log := range vm.Logs() {
+		statelog := evm.NewLog(log.Address, log.Topics, log.Data, header.Number.Uint64())
+		statedb.AddLog(statelog)
+	}
+	usedGas := vm.UsedGas()
+	totalUsedGas.Add(totalUsedGas, usedGas)
+
+	receipt := types.NewReceipt(statedb.IntermediateRoot().Bytes(), totalUsedGas)
+	receipt.TxHash = tx.Hash()
+	receipt.GasUsed = new(big.Int).Set(totalUsedGas)
+	if MessageCreatesContract(tx) {
+		from, _ := tx.From()
+		receipt.ContractAddress = crypto.CreateAddress(from, tx.Nonce())
+	}
+
+	logs := statedb.GetLogs(tx.Hash())
+	receipt.Logs = logs
+	receipt.Bloom = types.CreateBloom(types.Receipts{receipt})
+
+	glog.V(logger.Debug).Infoln(receipt)
+
+	vm.Free()
```

Comment:
Would it save any performance to bump `vm.Free()` up to after where it is last used in the function (line 144)? I like the "readability" of having it at the bottom, but just a thought.

Reply:
I don't think so. It would just mean that VM is freed early.

The problem is that Go only execute one transaction at a time, so it's always one VM created, one VM freed. We don't save any time by moving it up.

Reply:
roger that

Reply:
and we can't just slap the whole thing into a go routine because there might be conflicts in state, yes?

Reply:
Won't be any conflict. Each VM executes independently and they never share any memory. But if indeed we try to execute multiple VMs in parallel, moving it up might give us a really small performance improvement.

However, right now I prefer readability. Besides I don't think this is the performance bottleneck. We can always modify it easily in a later PR.